module4/lessons/Week4/SecurityConcerns.md

```diff
@@ -0,0 +1,187 @@
+
+```
+"ConnectionStrings": {
+  "GoodBooksMvcDb": "Server=localhost;Database=GoodBooksMvc;Port=5432;Username=postgres;Password=password123"
+}
+```
+
+We _could_ keep track of all of our connection strings in the application, but that opens us up to security risks.  Because our application is deployed on the web, we do not want to record the database configuration anywhere public (like github).  So, we can use **environment variables** instead.
+
+Environment Variables are stored globally on our machines.  Once an environment variable is created, it is available from nearly anywhere on our computer; this means that multiple programs could use the same environment variables.
+
+With an [environment variable set up](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_environment_variables?view=powershell-7.3), we can remove the connection string configuration in our appsettings.json, and get the connection string directly from the environment.
+
+```c#
+builder.Services.AddDbContext<GoodBooksMvcContext>(options =>
+    options.UseNpgsql(Environment.GetEnvironmentVariable("GOODBOOKS_DBCONNECTIONSTRING")).UseSnakeCaseNamingConvention());
+```
+
+**Warning** After creating environment variables, you will want to re-start visual studio before running any database commands!
+
+### Secrets Management
+
+Another solution to the problem of needing sensitive data for our applications to run is **Secrets Management**.  Similar to environment variables, this strategy allows us to move the storage of sensitive data _outside_ of our solution; outside of the range of our git tracking!
+
+In Visual Studio, managing secrets is fairly straightforward.  You can add project specific secrets in the project menu:
+
+<img src="/assets/images/module4/week4/ManageUserSecrets.png" style="width: 80%;">
+
+This will open a new file - secrets.json.  Here, you can create key/value pairs for data that you need to keep ... secret ...
+
+```json
+// secrets.json
+{
+  "GOODBOOKS_DBCONNECTIONSTRING": "Server=localhost;Database=GoodBooksMvc;Port=5432;Username=postgres;Password=password123"
+}
+```
+
+Using the Secrets Manager, you can call keys directly from `builder.Configuration`.
+
+```c#
+builder.Services.AddDbContext<GoodBooksMvcContext>(options =>
+  options
+    .UseNpgsql(builder.Configuration["GOODBOOKS_DBCONNECTIONSTRING"]
+    .UseSnakeCaseNamingConvention());
+```
+
+The secrets.json file is stored on your local machine, outside of the project file structure.  In your .csproj file, there will now be a unique identifier that the project uses to locate the file.
+
+```
+// GoodBooksMvc.csproj
+
+  <PropertyGroup>
+    ...
+    <UserSecretsId>e20e7f2a-cb30-4dd4-8139-ed66697eb566</UserSecretsId>
+  </PropertyGroup>
+```
+
```

Comment:
Might be interesting to have a question here about what happens if you change a secret value while the application is running? And then try it. I haven't tried this out in .NET but I'm pretty sure the secrets are always only read in when the application starts and this causes quite a bit of confusion for students.

Reply:
Oh, interesting... I hadn't even thought of that - to me secrets are kind of 'set it and forget it'; I'm curious what students are doing to run into this issue?  Have you had them change connection strings on the fly?